src/MC/fix_hmc.cpp

```diff
@@ -0,0 +1,1185 @@
+  if (angle_flag) rev_comm[m++] = force->newton_bond;
+  if (dihedral_flag) rev_comm[m++] = force->newton_bond;
+  if (improper_flag) rev_comm[m++] = force->newton_bond;
+  if (kspace_flag) rev_comm[m++] = force->kspace->tip4pflag;
+  for (i = ne; i < nv; i++) rev_comm[m++] = 0;
+
+  // Initialize array of pointers to manage per-atom energies:
+  delete[] eatomptr;
+  eatomptr = new double **[ne];
+  m = 0;
+  if (pair_flag) eatomptr[m++] = &force->pair->eatom;
+  if (bond_flag) eatomptr[m++] = &force->bond->eatom;
+  if (angle_flag) eatomptr[m++] = &force->angle->eatom;
+  if (dihedral_flag) eatomptr[m++] = &force->dihedral->eatom;
+  if (improper_flag) eatomptr[m++] = &force->improper->eatom;
+  if (kspace_flag) eatomptr[m++] = &force->kspace->eatom;
+
+  // Initialize array of pointers to manage per-atom virials:
+  delete[] vatomptr;
+  vatomptr = new double ***[nv];
+  m = 0;
+  if (pair_flag) vatomptr[m++] = &force->pair->vatom;
+  if (bond_flag) vatomptr[m++] = &force->bond->vatom;
+  if (angle_flag) vatomptr[m++] = &force->angle->vatom;
+  if (dihedral_flag) vatomptr[m++] = &force->dihedral->vatom;
+  if (improper_flag) vatomptr[m++] = &force->improper->vatom;
+  if (kspace_flag) vatomptr[m++] = &force->kspace->vatom;
+  for (i = 0; i < modify->nfix; i++)
+    if (modify->fix[i]->virial_peratom_flag) vatomptr[m++] = &modify->fix[i]->vatom;
+
+  // Determine the maximum and the actual numbers of per-atom variables in reverse
+  // communications:
+  // Note: fix-related virials do not communicate (thus 'ne' used instead of 'nv')
+  comm_reverse = 0;
+  ncommrev = 0;
+  for (m = 0; m < ne; m++)
+    if (rev_comm[m]) {
+      comm_reverse += 7;    // 1 energy + 6 virials
+      if (peatom_flag) ncommrev++;
+      if (pressatom_flag) ncommrev += 6;
+    }
+
+  // Determine maximum number of per-atom variables in forward and reverse
+  // communications when dealing with rigid bodies:
+  if (rigid_flag) {
+    comm_reverse = MAX(comm_reverse, 6);
+    comm_forward = 12;
+  }
+}
+
+/* ---------------------------------------------------------------------- */
+
+int FixHMC::setmask()
+{
+  int mask = 0;
+  mask |= END_OF_STEP;
+  return mask;
+}
+
+/* ---------------------------------------------------------------------- */
+
+void FixHMC::init()
+{
+  int ntimestep = update->ntimestep;
+
+  // Check whether there is any fixes that change box size and/or shape:
+  for (int i = 0; i < modify->nfix; i++) {
+    if (modify->fix[i]->box_change)
+      error->all(FLERR, "fix hmc is incompatible with fixes that change box size or shape");
+  }
+
+  // Check whether there are subsequent fixes with active virial_flag:
+  int first = modify->find_fix(this->id) + 1;
+  if (rigid_flag) first++;
+  for (int i = first; i < modify->nfix; i++)
+    if (modify->fix[i]->virial_peratom_flag || modify->fix[i]->virial_global_flag) {
+      if (comm->me == 0) printf("Fix %s defined after fix hmc.\n", modify->fix[i]->style);
```

Comment:
Please don't use printf(), but `utils::logmesg()` for any kind of output, so the redirection to files for out works properly. The function supports printf() style formatted strings but uses python-like fmt::format() syntax with '{}' instead of '%s'.